File: sdk/cocos/QaSDK.ts
// QaSDK — Cocos Creator (TypeScript) SDK 主入口
// 上报地址约定：POST {baseUrl}/api/qa/report
//
// 完整协议（与 FileApiLogger.cs 模板对齐）：
//   - 用例结果：report(name, result, message, tags)
//   - 自由日志：log(message, tags)
//   - API 拦截事件：logRequest / logResponse / logError
//       （对应 FileApiLogger 的 REQUEST / RESPONSE / ERROR 三类 gRPC 拦截事件）
//   - 任意原始上报：sendRaw(payload)（可直接转发 FileApiLogger 的 JSONL 行）
// 上报前会对 request / response / headers 中的敏感字段自动脱敏。
import { QaConfig } from './QaConfig'

let initialized = false

// 敏感字段（落库前脱敏，对应 FileApiLogger.SensitiveFieldNames）
const SENSITIVE_KEYS = new Set([
  'credential', 'authtoken', 'token', 'password', 'secret', 'apikey', 'key', 'authorization',
])

export function init(baseUrl?: string, token?: string, enabled?: boolean): void {
  if (baseUrl !== undefined) QaConfig.baseUrl = baseUrl
  if (token !== undefined) QaConfig.token = token
  if (enabled !== undefined) QaConfig.enabled = enabled
  initialized = true
}

// 递归脱敏：把对象中敏感字段的值替换为 ***
function redact(value: any): any {
  if (Array.isArray(value)) return value.map(redact)
  if (value && typeof value === 'object') {
    const out: Record<string, unknown> = {}
    for (const k of Object.keys(value)) {
      if (SENSITIVE_KEYS.has(k.toLowerCase())) out[k] = '***'
      else out[k] = redact(value[k])
    }
    return out
  }
  return value
}

interface ReportPayload {
  event: string
  name: string
  result: string
  message: string
  tags: Record<string, unknown>
  timestamp: number
  type?: string
  method?: string
  seq?: number
  headers?: Record<string, unknown>
  request?: unknown
  response?: unknown
  error?: string
  elapsed_ms?: number
}

function buildPayload(name: string, result: string, message: string, tags: Record<string, unknown>, event: string): ReportPayload {
  return {
    event: event || 'case_result',
    name: name || '',
    result: result || 'passed',
    message: message || '',
    tags: tags || {},
    timestamp: Date.now(),
  }
}

async function send(payload: ReportPayload): Promise<boolean> {
  if (!QaConfig.enabled) return false
  if (!initialized) init()
  const url = QaConfig.baseUrl.replace(/\/$/, '') + '/api/qa/report'
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  if (QaConfig.token) headers['Authorization'] = 'Bearer ' + QaConfig.token
  // 脱敏：request / response / headers 中的敏感字段
  const safe: ReportPayload = { ...payload }
  if (safe.headers) safe.headers = redact(safe.headers)
  if (safe.request !== undefined) safe.request = redact(safe.request)
  if (safe.response !== undefined) safe.response = redact(safe.response)
  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(safe),
    })
    return resp.ok
  } catch (e) {
    console.error('[QaSDK] 上报失败:', (e as Error).message)
    return false
  }
}

export function report(name: string, result = 'passed', message = '', tags: Record<string, unknown> = {}): Promise<boolean> {
  return send(buildPayload(name, result, message, tags, 'case_result'))
}

export function log(message: string, tags: Record<string, unknown> = {}): Promise<boolean> {
  return send(buildPayload('log', 'info', message, tags, 'log'))
}

export function logRequest(method: string, headers: Record<string, unknown> = {}, request?: unknown, opts: { seq?: number; elapsedMs?: number; tags?: Record<string, unknown>; message?: string } = {}): Promise<boolean> {
  return send({
    event: 'request', type: 'REQUEST', name: method || '', method: method || '',
    headers, request, seq: opts.seq || 0, elapsed_ms: opts.elapsedMs || 0,
    tags: opts.tags || {}, message: opts.message || '', timestamp: Date.now(),
  })
}

export function logResponse(method: string, headers: Record<string, unknown> = {}, response?: unknown, elapsedMs = 0, opts: { seq?: number; tags?: Record<string, unknown>; message?: string } = {}): Promise<boolean> {
  return send({
    event: 'response', type: 'RESPONSE', name: method || '', method: method || '',
    headers, response, elapsed_ms: elapsedMs, seq: opts.seq || 0,
    tags: opts.tags || {}, message: opts.message || '', timestamp: Date.now(),
  })
}

export function logError(method: string, error: string, elapsedMs = 0, opts: { headers?: Record<string, unknown>; seq?: number; tags?: Record<string, unknown> } = {}): Promise<boolean> {
  return send({
    event: 'error', type: 'ERROR', name: method || '', method: method || '',
    headers: opts.headers || {}, error: error || '', elapsed_ms: elapsedMs,
    seq: opts.seq || 0, tags: opts.tags || {}, message: error || '', timestamp: Date.now(),
  })
}

export function sendRaw(payload: ReportPayload): Promise<boolean> {
  return send(payload)
}

export const QaSDK = { init, report, log, logRequest, logResponse, logError, sendRaw }
export default QaSDK

File: sdk/python/qa_sdk.py
# QaSDK — Python SDK 主入口
# 上报地址约定：POST {BASE_URL}/api/qa/report
#
# 完整协议（与 FileApiLogger.cs 模板对齐）：
#   - 用例结果：report(name, result, message, tags)
#   - 自由日志：log(message, tags)
#   - API 拦截事件：log_request / log_response / log_error
#       （对应 FileApiLogger 的 REQUEST / RESPONSE / ERROR 三类 gRPC 拦截事件）
#   - 任意原始上报：send_raw(payload)（可直接转发 FileApiLogger 的 JSONL 行）
# 上报前会对 request / response / headers 中的敏感字段自动脱敏。
# 依赖：pip install requests
import json
import time

import requests

import qa_config as cfg

_initialized = False

# 敏感字段（落库前脱敏，对应 FileApiLogger.SensitiveFieldNames）
SENSITIVE_KEYS = {"credential", "authtoken", "token", "password", "secret", "apikey", "key", "authorization"}


def init(base_url=None, token=None, enabled=None):
    global _initialized
    if base_url is not None:
        cfg.BASE_URL = base_url
    if token is not None:
        cfg.TOKEN = token
    if enabled is not None:
        cfg.ENABLED = enabled
    _initialized = True


def _redact(value):
    """递归脱敏：把对象中敏感字段的值替换为 ***。"""
    if isinstance(value, list):
        return [_redact(v) for v in value]
    if isinstance(value, dict):
        return {k: ("***" if k.lower() in SENSITIVE_KEYS else _redact(v)) for k, v in value.items()}
    return value


def _build_payload(name, result, message, tags, event="case_result"):
    return {
        "event": event,
        "name": name or "",
        "result": result or "passed",
        "message": message or "",
        "tags": tags or {},
        "timestamp": int(time.time() * 1000),
    }


def _send(payload):
    if not cfg.ENABLED:
        return False
    if not _initialized:
        init()
    # 脱敏：request / response / headers 中的敏感字段
    safe = dict(payload)
    if "headers" in safe:
        safe["headers"] = _redact(safe.get("headers"))
    if "request" in safe:
        safe["request"] = _redact(safe.get("request"))
    if "response" in safe:
        safe["response"] = _redact(safe.get("response"))
    url = cfg.BASE_URL.rstrip("/") + "/api/qa/report"
    headers = {"Content-Type": "application/json"}
    if cfg.TOKEN:
        headers["Authorization"] = "Bearer " + cfg.TOKEN
    try:
        resp = requests.post(url, json=safe, headers=headers, timeout=10)
        return resp.ok
    except Exception as e:  # 上报失败不应阻断主流程
        print("[QaSDK] 上报失败:", e)
        return False


def report(name, result="passed", message="", tags=None):
    """上报一条用例结果。"""
    return _send(_build_payload(name, result, message, tags, "case_result"))


def log(message, tags=None):
    """上报一条日志。"""
    return _send(_build_payload("log", "info", message, tags, "log"))


def log_request(method, headers=None, request=None, seq=0, tags=None, message=""):
    """上报一次 API 请求（对应 FileApiLogger REQUEST）。"""
    return _send({
        "event": "request", "type": "REQUEST", "name": method or "",
        "method": method or "", "headers": headers or {}, "request": request,
        "seq": seq, "tags": tags or {}, "message": message,
        "timestamp": int(time.time() * 1000),
    })


def log_response(method, headers=None, response=None, elapsed_ms=0.0, seq=0, tags=None, message=""):
    """上报一次 API 响应（对应 FileApiLogger RESPONSE）。"""
    return _send({
        "event": "response", "type": "RESPONSE", "name": method or "",
        "method": method or "", "headers": headers or {}, "response": response,
        "elapsed_ms": elapsed_ms, "seq": seq, "tags": tags or {},
        "message": message, "timestamp": int(time.time() * 1000),
    })


def log_error(method, error="", elapsed_ms=0.0, headers=None, tags=None, seq=0):
    """上报一次 API 错误（对应 FileApiLogger ERROR）。"""
    return _send({
        "event": "error", "type": "ERROR", "name": method or "",
        "method": method or "", "headers": headers or {}, "error": error or "",
        "elapsed_ms": elapsed_ms, "seq": seq, "tags": tags or {},
        "message": error or "", "timestamp": int(time.time() * 1000),
    })


def send_raw(payload):
    """任意原始上报（可直接转发 FileApiLogger 的 JSONL 行）。"""
    return _send(payload)


if __name__ == "__main__":
    QaSDK = __import__("qa_sdk")
    QaSDK.init(cfg.BASE_URL, cfg.TOKEN, cfg.ENABLED)
    QaSDK.report("示例用例", "passed", "运行成功")
